core/src/main/java/org/jruby/RubyDir.java

```diff
@@ -206,17 +207,37 @@ private static ByteList globArgumentAsByteList(ThreadContext context, IRubyObjec
      * with each filename is passed to the block in turn. In this case, Nil is
      * returned.
      */
-    @JRubyMethod(required = 1, optional = 1, meta = true)
+    @JRubyMethod(required = 1, optional = 2, meta = true)
     public static IRubyObject glob(ThreadContext context, IRubyObject recv, IRubyObject[] args, Block block) {
         Ruby runtime = context.runtime;
-        int flags = args.length == 2 ? RubyNumeric.num2int(args[1]) : 0;
+        int flags = 0;
+        String dir, base = "";
+
+        if(args.length == 3) {
+            RubyHash hash = (RubyHash) args[2];
+            base = (String) hash.get(context.runtime.newSymbol("base"));
+            flags = RubyNumeric.num2int(args[1]);
+        } else if(args.length == 2) {
+            IRubyObject tmp = TypeConverter.checkHashType(runtime, args[1]);
+            if(tmp.isNil()){
+              flags = RubyNumeric.num2int(args[1]);
+            } else {
+              RubyHash hash = (RubyHash) args[1];
+              base = (String) hash.get(context.runtime.newSymbol("base"));
+            }
+        }
 
         List<ByteList> dirs;
         IRubyObject tmp = args[0].checkArrayType();
-        if (tmp.isNil()) {
-            dirs = Dir.push_glob(runtime, runtime.getCurrentDirectory(), globArgumentAsByteList(context, args[0]), flags);
+
+        if (!base.isEmpty() && !(new File(base).exists())){
+            dirs = new ArrayList<ByteList>();
+        } else if (tmp.isNil()) {
+            dir = base.isEmpty() ? runtime.getCurrentDirectory() : base;
+            dirs = Dir.push_glob(runtime, dir, globArgumentAsByteList(context, args[0]), flags);
```

Comment:
Wondering what the difference between ``runtime.getCurrentDirectory()`` and ``getCWD(runtime)`` or is it possible to simplify the code and only use either or?

Reply:
@ChrisBr getCWD will force canonicalization whereas runtime.getCurrentDirectory is whatever it was set to (which might have already been canonicalized).  Even if we happen to always canon. the stored value getCWD always tries to do it again (unless it is an URI).

Dir.glob is a dirty beast and I do not remember what contract it has for the incoming directory but that would be the proper guidance on which to use.